pennylane/devices/qubit/preprocess.py

```diff
@@ -138,8 +136,27 @@ def validate_and_expand_adjoint(
                 "to silence this warning.",
                 UserWarning,
             )
-        else:
-            trainable_params.append(k)
+
+    # determine the trainable params of the expanded tape by examining the decompositions
+    # of operators corresponding to trainable params of the old tape
+
+    trainable_params = []
+    old_offset = len(circuit._prep)
+    new_offset = old_offset
+    for op, decomp in zip(circuit._ops, decompositions):
+        num_decomp_params = sum(len(sub_op.data) for sub_op in decomp)
+        trainable = [i for i in range(len(op.data)) if i + old_offset in circuit.trainable_params]
+
+        if len(op.data) == num_decomp_params:
+            trainable_params.extend(i + new_offset for i in trainable)
+        elif trainable:
+            # let every new param be trainable if at least one old param is trainable.
+            # this works around cases such as U2 where we might not know which parameters
+            # of the decomposition corresponds to which parameters of the original op.
```

Comment:
perhaps we should add a warning here? something like:
```python
if len(op.data) < len(trainable):
    warnings.warn(f"some non-trainable parameters are being treated as trainable because of ambiguity from decomposition for {op.name}", UserWarning)